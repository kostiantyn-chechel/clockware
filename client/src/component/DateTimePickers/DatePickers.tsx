import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import { today } from '../../helpers/dateTime';

const useStyles = makeStyles((theme) => ({
    container: {
        display: 'flex',
        flexWrap: 'wrap',
    },
    textField: {
        marginLeft: theme.spacing(1),
        marginRight: theme.spacing(1),
        width: 200,
    },
}));

interface IDatePickers {
    onChange(event: React.ChangeEvent<{ value: string }>): void;
    defaultDate?: string;
}

const DatePickers:React.FC<IDatePickers> = (props) => {
    const classes = useStyles();

    const defDate = () => {
        if (props.defaultDate) return props.defaultDate;
        return today();
    };

    return (
        <div className={classes.container} >
            <TextField
                onChange={props.onChange}
                id="date"
                label="Дата:"
                type="date"
                // defaultValue={today()}
                defaultValue={defDate()}
                className={classes.textField}
                InputLabelProps={{
                    shrink: true,
                }}
            />
        </div>
    );
};

export default DatePickers;