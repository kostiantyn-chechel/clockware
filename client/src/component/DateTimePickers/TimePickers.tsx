import * as React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import { nowTimePlus } from "../../helpers/dateTime";

const useStyles = makeStyles((theme) => ({
    container: {
        display: 'flex',
        flexWrap: 'wrap',
    },
    textField: {
        marginLeft: theme.spacing(1),
        marginRight: theme.spacing(1),
        width: 200,
    },
}));

interface ItimePickers {
    onChange(event: React.ChangeEvent<{ value: string; }>): void;
}

const TimePickers: React.FC<ItimePickers> = (props) => {
    const classes = useStyles();

    return (
        <div className={classes.container}>
            <TextField
                onChange={props.onChange}
                id="time"
                label="Время:"
                type="time"
                defaultValue={nowTimePlus()}
                className={classes.textField}
                InputLabelProps={{
                    shrink: true,
                }}
                inputProps={{
                    step: 3600, // 60 min
                }}
            />
        </div>
    );
};

export default TimePickers;