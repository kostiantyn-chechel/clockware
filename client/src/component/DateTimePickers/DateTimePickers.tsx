import React, { useState, useEffect } from 'react';
import FormLabel from "@material-ui/core/FormLabel";
import Grid from "@material-ui/core/Grid";
import DatePickers from "./DatePickers";
import TimePickers from "./TimePickers";
import makeStyles from "@material-ui/core/styles/makeStyles";
import Warning from "../Warning";
import { dateToString } from "../../helpers/dateTime";

const useStyles = makeStyles((theme) => ({
    time: {
        display: 'flex',
        justifyContent: 'space-around',
    },
    label: {
        margin: theme.spacing(1, 0, 0, 2),
    },
}));

interface IDateTimePickers {
    handleSelectDate(date: string): void,
    handleSelectTime(time: string): void,
}

const DateTimePickers: React.FC<IDateTimePickers> = (props) => {
    const classes = useStyles();

    const today = new Date();
    const nowTime = today.getHours() + 1;
    today.setHours(nowTime,0,0);


    const [choiceDataTime, setChoiceDataTime] = useState<Date>(today);
    const [valid, setValid] = useState(true);

    /* eslint-disable */
    useEffect(() => {
        if (choiceDataTime >= today) {
            setValid(true);
            props.handleSelectDate(dateToString(choiceDataTime));
            props.handleSelectTime(choiceDataTime.getHours() + ':00')
        } else {
            setValid(false);
        }
    },[choiceDataTime]);
    /* eslint-enable */

    const selectDay = (event: React.ChangeEvent<{ value: string; }>) => {
        event.preventDefault();
        const choiceDay = new Date(event.target.value);
        choiceDay.setHours(choiceDataTime.getHours(),0,0);
        setChoiceDataTime(choiceDay);
    };

    const selectTime = (event: React.ChangeEvent<{ value: string; }>) => {
        event.preventDefault();
        const choiceDay = new Date(choiceDataTime);
        choiceDay.setHours(+event.target.value.slice(0,2));
        setChoiceDataTime(choiceDay);
    };

    return (
        <React.Fragment>
            <FormLabel className={classes.label} component="legend">Укажите дату и время:</FormLabel>
            <Grid className={classes.time} item xs={12}>
                <Grid item xs={12} sm={6}>
                    <DatePickers
                        onChange={selectDay}
                    />
                </Grid>
                <Grid item xs={12} sm={6}>
                    <TimePickers
                        onChange={selectTime}
                    />
                </Grid>
            </Grid>
            <Warning valid={valid}>
                Выбранные дата и время уже прошли!
            </Warning>
        </React.Fragment>
    );
};

export default DateTimePickers;