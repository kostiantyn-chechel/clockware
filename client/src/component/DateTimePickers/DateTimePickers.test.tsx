import React from 'react';
import { shallow, configure, mount } from 'enzyme';
import renderer from 'react-test-renderer'
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

import DateTimePickers from './DateTimePickers';

const NEW_DATA = '2021-02-02';
const EARLY_DATA = '2021-01-02';

describe('Test <DateTimePickers>', () => {
    const handleSelectDateMock = jest.fn();
    const handleSelectTimeMock = jest.fn();

    const Component = <DateTimePickers
        handleSelectDate={handleSelectDateMock}
        handleSelectTime={handleSelectTimeMock}
    />;

    const ShallowComponent = shallow(Component);

    test('DateTimePickers component should exists', () => {
        expect(ShallowComponent).toBeDefined();
    });

    test("ToMatchSnapshot DateTimePickers with default date", () => {
        expect(renderer.create(Component).toJSON()).toMatchSnapshot();
    });

    test('Changes value after change Data', () => {
        const MountComponent = mount(Component);

        MountComponent.find('#date-label').first().simulate('change', {
            target: { value: NEW_DATA },
        });

        expect(handleSelectDateMock).toHaveBeenCalledTimes(1);

    });

    describe('Date and time earlier than current', () => {

        test('Error message expected', () => {
            const MountComponent = mount(Component);

            MountComponent.find('#date-label').first().simulate('change', {
                target: { value: EARLY_DATA },
            });

            const node = MountComponent.find('.makeStyles-warningBlock-43');

            expect(node).toBeDefined();

            // expect(node).toHaveLength(1);

        });
    });

});
