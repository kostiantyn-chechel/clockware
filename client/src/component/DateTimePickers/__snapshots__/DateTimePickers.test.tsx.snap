// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <DateTimePickers> ToMatchSnapshot DateTimePickers with default date 1`] = `
Array [
  <legend
    className="MuiFormLabel-root makeStyles-label-2"
  >
    Укажите дату и время:
  </legend>,
  <div
    className="MuiGrid-root makeStyles-time-1 MuiGrid-item MuiGrid-grid-xs-12"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
    >
      <div
        className="makeStyles-container-3"
      >
        <div
          className="MuiFormControl-root MuiTextField-root makeStyles-textField-4"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
            data-shrink={true}
            htmlFor="date"
            id="date-label"
          >
            Дата:
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              defaultValue="2021-03-10"
              disabled={false}
              id="date"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="date"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
    >
      <div
        className="makeStyles-container-5"
      >
        <div
          className="MuiFormControl-root MuiTextField-root makeStyles-textField-6"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
            data-shrink={true}
            htmlFor="time"
            id="time-label"
          >
            Время:
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              defaultValue="16:00"
              disabled={false}
              id="time"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              step={3600}
              type="time"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
